imsim/photon_pooling.py

```diff
@@ -0,0 +1,547 @@
+                del photons  # As with the stamps above, let the garbage collector know we don't need the photons anymore.
+
+                # Note: in typical imsim usage, all current_vars will be 0. So this normally doesn't
+                # add much to the checkpointing data.
+                nz_var = np.nonzero(current_vars)[0]
+                all_vars.extend([current_vars[k] for k in nz_var])
+
+            if self.checkpoint is not None:
+                self.save_checkpoint(self.checkpoint, chk_name, base, full_image, all_stamps, all_vars, all_obj_nums, batch_num+1)
+
+        # Bring the image so far up to a flat noise variance
+        current_var = galsim.config.FlattenNoiseVariance(
+                base, full_image, all_stamps, tuple(all_vars), logger)
+
+        return full_image, current_var
+
+    @staticmethod
+    def merge_photon_arrays(stamps):
+        """Given a list of stamps of photon objects, return a single merged photon array.
+
+        Parameters:
+            stamps: list of galsim.PhotonArrays to be merged into one.
+
+        Returns:
+            merged: A single PhotonArray containing all the photons.
+        """
+        n_tot = sum(len(stamp.photons) for stamp in stamps)
+        merged = galsim.PhotonArray(n_tot)
+        start = 0
+        for stamp in stamps:
+            merged.copyFrom(stamp.photons, slice(start, start+stamp.photons.size()))
+            start += len(stamp.photons)
+        return merged
+
+    @staticmethod
+    def accumulate_photons(photons, imview, sensor, resume=False, recalc=True):
+        """Accumulate a photon array onto a sensor.
+
+        Parameters:
+            photons: A PhotonArray containing the photons to be accumulated.
+            imview: The image view to which we draw the accumulated photons.
+            sensor: Sensor to use for accumulation.
+            resume: Resume accumulating following an earlier call for some extra performance. Default False.
+            recalc: Recalculate the pixel boundaries for the sensor. Default True.
+        """
+        # Below we check whether sensor is a SiliconSensor. If it is, use the
+        # recalc argument to update pixel boundaries at the start of the call.
+        # Regular Sensors don't simulate this so don't accept recalc.
+        if imview.dtype in (np.float32, np.float64):
+            if isinstance(sensor, SiliconSensor):
+                sensor.accumulate(photons, imview, imview.center, resume=resume, recalc=recalc)
+            else:
+                sensor.accumulate(photons, imview, imview.center, resume=resume)
+        else:
+            # Create a temporary ImageD to work in.
+            im1 = galsim.image.ImageD(bounds=imview.bounds)
+            if isinstance(sensor, SiliconSensor):
+                sensor.accumulate(photons, im1, imview.center, resume=resume, recalc=True)
```

Comment:
I think recalc and resume won't work here, since it's always a new image.  We probably don't ever use this branch in imsim, but just in case, maybe issue a warning if resume is True or recalc is False that these don't work when the image type is not float32 or float64.

Reply:
Thanks for the suggestion -- adding warnings now.